### PR TITLE
V2 refactor

### PR DESCRIPTION
**Pull request by Deivu**
This is the WIP update for Shoukaku, this aims to fix and reduce the clutter of v1 while providing the same benefits of v1

Progress:
[x] Code
[x] Testing
[x] Documentation
[x] Typescript Typings

For discussions or recommendations, feel free to comment down in this thread

### CODE REVIEW

**Review by Kyosodev — APPROVED** [review]

**Comment by inhydrox** [comment]
Some fixes:
1. https://github.com/Deivu/Shoukaku/blob/next/src/guild/ShoukakuConnection.js#L140 - It doesn't work
2. https://github.com/Deivu/Shoukaku/blob/next/src/node/ShoukakuSocket.js#L51 - Returns 0, even the bot is connected

**Comment by inhydrox** [comment]
Yeah and maybe don't forget this https://github.com/Deivu/Shoukaku/issues/38

**Comment by Gaeta** [comment]
Any ETA on v2 refactor?

**Comment by Deivu** [comment]
No ETA for now, most likely when Lavalink dev goes to stable, and Discord.JS v13 gets finalized in its breaking changes, though it also depends on how fast I work, or when I will resume coding again

**Comment by Deivu** [comment]
@inhydrox state should now be turning to connected in voice connection, as for moveChannel, I dropped it as I don't think I need to support that feature with Shoukaku as this deals with Discord REST which is not something I want to support with Shoukaku

**Comment by hydrox19** [comment]
> @inhydrox state should now be turning to connected in voice connection, as for moveChannel, I dropped it as I don't think I need to support that feature with Shoukaku as this deals with Discord REST which is not something I want to support with Shoukaku

Alright, I'll test the voice connection one and let you know if I found any more bugs.

**Comment by kakarot-dev** [comment]
Hmmm or ded

**Comment by 0t4u** [comment]
Types for next are in PR (#51)

**Comment by 0t4u** [comment]
So um my computer borked don’t expect updates lmfao

**Comment by Deivu** [comment]
@0t4u  there have been changes, quite a bit so typings may need to change, just do it at your own pace, as the recent commits enables the lib to be used on other libraries with ease

**Comment by 0t4u** [comment]
@Deivu updated. Please check if it works.

**Comment by kakarot-dev** [comment]
I specifically saw this line before i asked if i can guildID to cameCase 
https://github.com/Deivu/Shoukaku/blob/8653a0c14bf6a294da10a8c8b3a127a0b6ad32b1/src/guild/ShoukakuConnection.js#L93

**Comment by kakarot-dev** [comment]
Why is the filter class having the volume option?
Wouldnt it be better if it was existed in the player class?

**Comment by 0t4u** [comment]
Lol the amount of commits

**Comment by kakarot-dev** [comment]
Is this pr ready for testing :octocat:

**Comment by kakarot-dev** [comment]
> Lol the amount of commits

Yea, sorry bout that, i used mobile and couldnt use my laptop xd
won’t happen again

**Comment by Deivu** [comment]
It could be tested, but there might be changes like changing things to camelcase one I decided to do so

**Comment by kakarot-dev** [comment]
Oh, can i do that
ill use laptop this time

**Comment by Deivu** [comment]
This is ready for testing and review, in order for this to get merged, First, We need to wait for Discord.JS v13 to be the current stable, and Secondly, This needs to be tested thoroughly to ensure this one won't cause issues, specially on bots in production.

**Comment by kakarot-dev** [comment]
Discord js v13 will be released when crawl wakes up. ( im not lying, he announced that in the announcement channel )

**Comment by Deivu** [comment]
> Discord js v13 will be released when crawl wakes up. ( im not lying, he announced that in the announcement channel )

The second one must be also be considered, which is a through test of the next refactor, I don't want bots breaking in production due to unforseen bugs in the library

**Comment by Orchit07** [comment]
> Discord js v13 will be released when crawl wakes up. ( im not lying, he announced that in the announcement channel )

and need node v16 
image

**Comment by Deivu** [comment]
Engine is updated

**Comment by kakarot-dev** [comment]
readme need to be fixed to show that node 16 is required

**Comment by kakarot-dev** [comment]
Types of ShoukakuTrack and ShoukakuTrackList arent there

**Comment by Deivu** [comment]
> Types of ShoukakuTrack and ShoukakuTrackList arent there
@0t4u Could you add this typings

**Comment by kakarot-dev** [comment]
if you want i could add them, i just need to know what “raw” will be

**Comment by 0t4u** [comment]
@Deivu gimme a few hrs

**Comment by Deivu** [comment]
> if you want i could add them, i just need to know what “raw” will be

Tracklist raw is this
```js
{
  "loadType": "TRACK_LOADED",
  "playlistInfo": {},
  "tracks": [
    {
      "track": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
      "info": {
        "identifier": "dQw4w9WgXcQ",
        "isSeekable": true,
        "author": "RickAstleyVEVO",
        "length": 212000,
        "isStream": false,
        "position": 0,
        "title": "Rick Astley - Never Gonna Give You Up",
        "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ"
      }
    }
  ]
}
```

ShoukakuTrack raw is the object inside the array of tracks

**Comment by kakarot-dev** [comment]
> > if you want i could add them, i just need to know what “raw” will be
> 
> Tracklist raw is this
> 
> ```js
> {
>   "loadType": "TRACK_LOADED",
>   "playlistInfo": {},
>   "tracks": [
>     {
>       "track": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
>       "info": {
>         "identifier": "dQw4w9WgXcQ",
>         "isSeekable": true,
>         "author": "RickAstleyVEVO",
>         "length": 212000,
>         "isStream": false,
>         "position": 0,
>         "title": "Rick Astley - Never Gonna Give You Up",
>         "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ"
>       }
>     }
>   ]
> }
> ```
> 
> ShoukakuTrack raw is the object inside the array of tracks

Ty ty

**Comment by 0t4u** [comment]
@Deivu @kakarot-dev it’s already in the Constants.d.ts file

**Comment by hydrox19** [comment]
Huh check the constants files in types folder

**Comment by hydrox19** [comment]
It is there

**Comment by Deivu** [comment]
> @Deivu @kakarot-dev it’s already in the Constants.d.ts file
smh I got jeibaited there

**Comment by kakarot-dev** [comment]
Oh, i got confused too when i didnt see the file xd

**Comment by hydrox19** [comment]
smh

**Comment by kakarot-dev** [comment]
One question 
tracklist has the property “playlistName”
while the example of @Deivu has “playlistInfo”
so…. Is that intentional?

**Comment by Deivu** [comment]
My structure don't have playlistInfo, instead its a property directly in the class for convenience 
https://github.com/Deivu/Shoukaku/blob/next/src/struct/ShoukakuTrackList.js

**Comment by kakarot-dev** [comment]
Oh, kk understood

**Comment by kakarot-dev** [comment]
Someone mentioned me?

**Comment by kakarot-dev** [comment]
Code tested in production and works perfectly.

File: types/Shoukaku.d.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    public library: GetterObj<any, any, any>;
+    public id?: Snowflake | null;
+    public nodes: Map<string, ShoukakuSocket>;
+    private options: Object;
+
+    public get players(): Map<Snowflake, ShoukakuPlayer>;
+
+    public addNode(options: NodeOptions): void;
+    public removeNode(name: string, reason?: string): void;
+    public getNode(query: string | Array<string>): ShoukakuSocket;
```
Comment (Orchit07) [inline types/Shoukaku.d.ts:24]: params `query` must be nullable

Comment (Deivu) [reply types/Shoukaku.d.ts]: Fixed

**Comment by Allvaa** [comment]
when setting volume to 0 using setFilters, its value will be always 1. because OR operator checks its boolean, and yeah 0 is false, so it will pick the 1.0 instead
https://github.com/Deivu/Shoukaku/blob/c45294fcc2966b114bbf8a9c1d37263f68c76656/src/struct/ShoukakuFilter.js#L51
maybe we can replace the OR operator above with nullish coalescing operator (??) ?

**Comment by kakarot-dev** [comment]
normally you would just use the mute method to remove voice, so yea....

**Comment by Allvaa** [comment]
> normally you would just use the mute method to remove voice, so yea....

indeed, but it's still a bug tho. the function doesn't set the correct value of what is given

**Comment by Deivu** [comment]
> when setting volume to 0 using setFilters, its value will be always 1. because OR operator checks its boolean, and yeah 0 is false, so it will pick the 1.0 instead
> https://github.com/Deivu/Shoukaku/blob/c45294fcc2966b114bbf8a9c1d37263f68c76656/src/struct/ShoukakuFilter.js#L51
> 
> 
> maybe we can replace the OR operator above with nullish coalescing operator (??) ?

Makes sense for it to be on ?? instead of ||, updated in latest commit

**Comment by Allvaa** [comment]
> > when setting volume to 0 using setFilters, its value will be always 1. because OR operator checks its boolean, and yeah 0 is false, so it will pick the 1.0 instead
> > https://github.com/Deivu/Shoukaku/blob/c45294fcc2966b114bbf8a9c1d37263f68c76656/src/struct/ShoukakuFilter.js#L51
> > 
> > maybe we can replace the OR operator above with nullish coalescing operator (??) ?
> 
> Makes sense for it to be on ?? instead of ||, updated in latest commit

thanks. anyway will sourceName be added to ShoukakuTrack#info?

**Comment by Deivu** [comment]
> thanks. anyway will sourceName be added to ShoukakuTrack#info?
Any relevant info on what git commit this was merged?

**Comment by Allvaa** [comment]
> > thanks. anyway will sourceName be added to ShoukakuTrack#info?
> 
> Any relevant info on what git commit this was merged?

you mean merged on dev branch? if so this is the commit https://github.com/freyacodes/Lavalink/commit/626c2186641d249e472a1070bdc61a99c077e6ba

File: index.d.ts [hunk 2]
```diff
@@ -1,434 +1,10 @@
...
-    tanOffset?: number;
-    tanScale?: number;
-    offset?: number;
-    scale?: number;
-}
+// idfk what this is lmao
 
+import { Shoukaku } from './types/Shoukaku';
+import * as Constants from './types/Constants';
+import { Libraries } from './types/libraries/Libraries';
```
Comment (HerrTxbias) [inline index.d.ts:5]:
```suggestion
import * as Libraries from './types/libraries/Libraries';
```

**Comment by Deivu** [comment]
I think this is mostly ironed out, feel free to comment if it's stable on your end or not

**Comment by kakarot-dev** [comment]
its stable for me 👍

File: types/guild/ShoukakuConnection.d.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  public state: state.CONNECTED | state.CONNECTING | state.DISCONNECTED | state.DISCONNECTING;
+  public moved: boolean;
+  public reconnecting: boolean;
+  private serverUpdate: boolean;
+
+  public setDeaf(deaf?: boolean): void;
+  public setMute(mute?: boolean): void;
+  public disconnect(): void;
+  public connect(options: ConnectOptions): Promise<void>;
+  public attemptReconnect(channelId: Snowflake): Promise<void>;
```
Comment (Orchit07) [inline types/guild/ShoukakuConnection.d.ts:31]: rename `attempReconnect` to `reconnect`

Comment (Deivu) [reply types/guild/ShoukakuConnection.d.ts:31]: oh forgot to rename that, could you pr the change for me, I'm eating right now

**Pull request merged**